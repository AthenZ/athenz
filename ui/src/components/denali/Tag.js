/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import PropTypes from 'prop-types';
import { css, cx } from 'emotion';
import { colors } from './styles/colors';
import { cssFontFamilies } from './styles/fonts';
import { rgba } from 'polished';
import Icon from '../denali/icons/Icon';
/**
 * Tags are used to display small pieces of information. There are three general
 * styles of tags: default, clickable, and removable.
 */
const makeCssTag = (props) => css`
    ${cssFontFamilies.default};
    background: ${props.disabled
        ? rgba(colors.black, 0.1)
        : props.clickable
        ? 'none'
        : 'rgba(53, 112, 244, 0.08)'};
    border-radius: ${props.small ? '11px' : '14px'};
    border: 1px solid ${props.clickable ? colors.link : 'transparent'};
    box-sizing: border-box;
    color: ${props.disabled
        ? rgba(colors.black, 0.2)
        : props.clickable
        ? colors.link
        : colors.black};
    cursor: ${props.disabled
        ? 'not-allowed'
        : props.clickable
        ? 'pointer'
        : undefined};
    display: inline-block;
    font-size: ${props.small ? '12px' : '14px'};
    height: ${props.small ? '22px' : '28px'};
    line-height: ${props.small ? '20px' : '26px'};
    margin: 3px;
    min-width: 50px;
    padding: 0 12px;
    position: relative;
    text-decoration: none;
    transition: ${!props.noanim ? 'all 0.2s ease-out' : undefined};
    vertical-align: top;
    white-space: nowrap;
    label: tag;
    &:hover {
        background: ${props.disabled ? rgba(colors.black, 0.1) : '#cbd9f9'};
    }
    & .tag-remove {
        align-items: center;
        background: #cbd9f9;
        border-radius: ${props.small ? '0 11px 11px 0' : '0 14px 14px 0'};
        box-shadow: none;
        box-sizing: border-box;
        display: flex;
        height: ${props.small ? '22px' : '28px'};
        justify-content: center;
        opacity: 0;
        padding-right: 2px;
        position: absolute;
        right: ${props.small ? '-11px' : '-14px'};
        top: -1px;
        transition: ${!props.noanim ? 'all 0.2s ease-out' : undefined};
        width: ${props.small ? '22px' : '28px'};
    }
    &:hover .tag-remove {
        box-shadow: -4px 0 5px -1px ${rgba(colors.black, 0.2)};
        opacity: 1;
        right: 0;
    }
`;

const Tag = (props) => {
    const {
        children,
        className,
        disabled,
        innerRef,
        noanim,
        small,
        onClick,
        onClickRemove,
    } = props;

    const cssProps = {
        disabled,
        noanim,
        small,
        clickable: Boolean(onClick),
    };
    const cssTag = makeCssTag(cssProps);

    return (
        <div
            className={cx(cssTag, 'denali-tag', className)}
            ref={innerRef}
            onClick={onClick}
            data-testid='tag'
        >
            {children}
            {onClickRemove && (
                <div className='tag-remove'>
                    <Icon
                        name='close'
                        color={rgba(colors.black, 0.6)}
                        colorHover={colors.black}
                        isLink
                        size={small ? '16px' : '20px'}
                        onClick={onClickRemove}
                    />
                </div>
            )}
        </div>
    );
};

Tag.propTypes = {
    /**
     * Children (usually text) to render inside tag
     * @ignore
     */
    children: PropTypes.any,
    /** Additonal class to apply to the outer div */
    className: PropTypes.string,
    /** Disable the tag */
    disabled: PropTypes.bool,
    /** Forward React ref to the outermost div of the component */
    innerRef: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
    /** Disable animations / transitions */
    noanim: PropTypes.bool,
    /** Toggle small tags */
    small: PropTypes.bool,
    /**
     * If an `onClick()` function is passed in, tag will be selectable.
     * **NOTE:** This is mutually exclusive with `onClickRemove()`
     */
    onClick: PropTypes.func,
    /**
     * If an `onClickRemove()` function is passed in, tag will have a remove button.
     * **NOTE:** This is mutually exclusivewith `onClick()`
     */
    onClickRemove: PropTypes.func,
    /**
     * Validator to ensure only one of `onClick()` / `onClickRemove()` is defined
     * @ignore
     */
    xorOnClick: function(props, propName, componentName) {
        if (props.onClick && props.onClickRemove) {
            return new Error(
                'Only one of `onClick` | `onClickRemove` canbe defined in `' +
                    componentName +
                    '`. Validation failed.'
            );
        }
    },
};

Tag.defaultProps = {
    disabled: false,
    noanim: false,
    small: false,
};

export default Tag;
