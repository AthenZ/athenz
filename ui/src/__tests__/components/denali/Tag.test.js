/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import Tag from '../../../components/denali/Tag';

describe('Tag', () => {
    it('should render', () => {
        const { getByTestId } = render(<Tag>A tag</Tag>);
        const tag = getByTestId('tag');

        expect(tag).toMatchSnapshot();
    });

    it('should render a removable Tag', () => {
        const onClickRemove = jest.fn();
        const { getByTestId } = render(
            <Tag onClickRemove={onClickRemove}>A tag</Tag>
        );
        const tag = getByTestId('tag');

        expect(tag).toMatchSnapshot();
        // fireEvent.click(tag.querySelector('.remove-button'));
        // expect(onClickRemove).toHaveBeenCalledTimes(1);
    });

    it('should render a selectable Tag', () => {
        const onClick = jest.fn();
        const { getByTestId } = render(<Tag onClick={onClick}>A tag</Tag>);
        const tag = getByTestId('tag');

        expect(tag).toMatchSnapshot();
        fireEvent.click(tag);
        expect(onClick).toHaveBeenCalledTimes(1);
    });

    it('should render a disabled Tag', () => {
        const { getByTestId } = render(<Tag disabled>A tag</Tag>);
        expect(getByTestId('tag')).toMatchSnapshot();
    });

    it('should render a small Tag', () => {
        const onClickRemove = jest.fn();
        const { getByTestId } = render(
            <Tag small onClickRemove={onClickRemove}>
                A tag
            </Tag>
        );

        expect(getByTestId('tag')).toMatchSnapshot();
    });

    it('should render with an additional class name', () => {
        const { getByTestId } = render(
            <Tag className='custom-class'>A tag</Tag>
        );
        const tag = getByTestId('tag');

        expect(tag).toHaveClass('denali-tag');
        expect(tag).toHaveClass('custom-class');
    });

    it('should disable transitions', () => {
        const { getByTestId } = render(<Tag noanim>A tag</Tag>);
        const tag = getByTestId('tag');

        expect(tag).toMatchSnapshot();
    });

    it('should warn on invalid use of onClick & onClickRemove', () => {
        const onClick = jest.fn();
        const onClickRemove = jest.fn();
        const consoleErrorSpy = jest
            .spyOn(console, 'error')
            .mockImplementation(() => {});

        render(
            <Tag onClick={onClick} onClickRemove={onClickRemove}>
                A tag
            </Tag>
        );

        expect(consoleErrorSpy).toHaveBeenCalledTimes(1);
    });
});
