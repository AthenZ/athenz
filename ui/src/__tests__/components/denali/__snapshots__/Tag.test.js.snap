// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag should disable transitions 1`] = `
.emotion-0 {
  font-family: Helvetica,Arial,sans-serif;
  background: rgba(53,112,244,0.08);
  border-radius: 14px;
  border: 1px solid transparent;
  box-sizing: border-box;
  color: #303030;
  display: inline-block;
  font-size: 14px;
  height: 28px;
  line-height: 26px;
  margin: 3px;
  min-width: 50px;
  padding: 0 12px;
  position: relative;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: top;
  white-space: nowrap;
}

.emotion-0:hover {
  background: #cbd9f9;
}

.emotion-0 .tag-remove {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #cbd9f9;
  border-radius: 0 14px 14px 0;
  box-shadow: none;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 28px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  opacity: 0;
  padding-right: 2px;
  position: absolute;
  right: -14px;
  top: -1px;
  width: 28px;
}

.emotion-0:hover .tag-remove {
  box-shadow: -4px 0 5px -1px rgba(48,48,48,0.2);
  opacity: 1;
  right: 0;
}

<div
  class="emotion-0 denali-tag"
  data-testid="tag"
>
  A tag
</div>
`;

exports[`Tag should render 1`] = `
.emotion-0 {
  font-family: Helvetica,Arial,sans-serif;
  background: rgba(53,112,244,0.08);
  border-radius: 14px;
  border: 1px solid transparent;
  box-sizing: border-box;
  color: #303030;
  display: inline-block;
  font-size: 14px;
  height: 28px;
  line-height: 26px;
  margin: 3px;
  min-width: 50px;
  padding: 0 12px;
  position: relative;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: all 0.2s ease-out;
  transition: all 0.2s ease-out;
  vertical-align: top;
  white-space: nowrap;
}

.emotion-0:hover {
  background: #cbd9f9;
}

.emotion-0 .tag-remove {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #cbd9f9;
  border-radius: 0 14px 14px 0;
  box-shadow: none;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 28px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  opacity: 0;
  padding-right: 2px;
  position: absolute;
  right: -14px;
  top: -1px;
  -webkit-transition: all 0.2s ease-out;
  transition: all 0.2s ease-out;
  width: 28px;
}

.emotion-0:hover .tag-remove {
  box-shadow: -4px 0 5px -1px rgba(48,48,48,0.2);
  opacity: 1;
  right: 0;
}

<div
  class="emotion-0 denali-tag"
  data-testid="tag"
>
  A tag
</div>
`;

exports[`Tag should render a disabled Tag 1`] = `
.emotion-0 {
  font-family: Helvetica,Arial,sans-serif;
  background: rgba(48,48,48,0.1);
  border-radius: 14px;
  border: 1px solid transparent;
  box-sizing: border-box;
  color: rgba(48,48,48,0.2);
  cursor: not-allowed;
  display: inline-block;
  font-size: 14px;
  height: 28px;
  line-height: 26px;
  margin: 3px;
  min-width: 50px;
  padding: 0 12px;
  position: relative;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: all 0.2s ease-out;
  transition: all 0.2s ease-out;
  vertical-align: top;
  white-space: nowrap;
}

.emotion-0:hover {
  background: rgba(48,48,48,0.1);
}

.emotion-0 .tag-remove {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #cbd9f9;
  border-radius: 0 14px 14px 0;
  box-shadow: none;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 28px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  opacity: 0;
  padding-right: 2px;
  position: absolute;
  right: -14px;
  top: -1px;
  -webkit-transition: all 0.2s ease-out;
  transition: all 0.2s ease-out;
  width: 28px;
}

.emotion-0:hover .tag-remove {
  box-shadow: -4px 0 5px -1px rgba(48,48,48,0.2);
  opacity: 1;
  right: 0;
}

<div
  class="emotion-0 denali-tag"
  data-testid="tag"
>
  A tag
</div>
`;

exports[`Tag should render a removable Tag 1`] = `
.emotion-1 {
  font-family: Helvetica,Arial,sans-serif;
  background: rgba(53,112,244,0.08);
  border-radius: 14px;
  border: 1px solid transparent;
  box-sizing: border-box;
  color: #303030;
  display: inline-block;
  font-size: 14px;
  height: 28px;
  line-height: 26px;
  margin: 3px;
  min-width: 50px;
  padding: 0 12px;
  position: relative;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: all 0.2s ease-out;
  transition: all 0.2s ease-out;
  vertical-align: top;
  white-space: nowrap;
}

.emotion-1:hover {
  background: #cbd9f9;
}

.emotion-1 .tag-remove {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #cbd9f9;
  border-radius: 0 14px 14px 0;
  box-shadow: none;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 28px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  opacity: 0;
  padding-right: 2px;
  position: absolute;
  right: -14px;
  top: -1px;
  -webkit-transition: all 0.2s ease-out;
  transition: all 0.2s ease-out;
  width: 28px;
}

.emotion-1:hover .tag-remove {
  box-shadow: -4px 0 5px -1px rgba(48,48,48,0.2);
  opacity: 1;
  right: 0;
}

.emotion-0 {
  fill: rgba(48,48,48,0.6);
  cursor: pointer;
  vertical-align: text-bottom;
}

<div
  class="emotion-1 denali-tag"
  data-testid="tag"
>
  A tag
  <div
    class="tag-remove"
  >
    <svg
      class="emotion-0"
      data-testid="icon"
      height="20px"
      viewBox="0 0 1024 1024"
      width="20px"
    >
      <title>
        x
      </title>
      <path
        d="M572.373 512l224.427-224.213c8.506-8.066 13.798-19.448 13.798-32.066 0-9.118-2.763-17.59-7.498-24.625l0.1 0.158c-7.858-10.565-20.305-17.335-34.333-17.335-11.534 0-21.999 4.577-29.678 12.013l0.011-0.011-227.2 225.707-226.133-225.707c-7.733-7.792-18.447-12.616-30.289-12.616-14.295 0-26.948 7.030-34.691 17.822l-0.087 0.128c-4.635 6.877-7.398 15.35-7.398 24.467 0 12.618 5.292 24 13.778 32.048l0.020 0.019 224.427 224.213-224.427 224.213c-8.506 8.066-13.798 19.448-13.798 32.066 0 9.118 2.763 17.59 7.498 24.625l-0.1-0.158c7.83 10.919 20.483 17.949 34.778 17.949 11.841 0 22.556-4.824 30.286-12.613l0.003-0.003 226.133-225.707 226.133 225.707c7.668 7.425 18.133 12.002 29.667 12.002 14.028 0 26.476-6.77 34.252-17.222l0.081-0.114c4.635-6.877 7.398-15.35 7.398-24.467 0-12.618-5.292-24-13.778-32.048l-0.020-0.019z"
      />
    </svg>
  </div>
</div>
`;

exports[`Tag should render a selectable Tag 1`] = `
.emotion-0 {
  font-family: Helvetica,Arial,sans-serif;
  background: none;
  border-radius: 14px;
  border: 1px solid #3697f2;
  box-sizing: border-box;
  color: #3697f2;
  cursor: pointer;
  display: inline-block;
  font-size: 14px;
  height: 28px;
  line-height: 26px;
  margin: 3px;
  min-width: 50px;
  padding: 0 12px;
  position: relative;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: all 0.2s ease-out;
  transition: all 0.2s ease-out;
  vertical-align: top;
  white-space: nowrap;
}

.emotion-0:hover {
  background: #cbd9f9;
}

.emotion-0 .tag-remove {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #cbd9f9;
  border-radius: 0 14px 14px 0;
  box-shadow: none;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 28px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  opacity: 0;
  padding-right: 2px;
  position: absolute;
  right: -14px;
  top: -1px;
  -webkit-transition: all 0.2s ease-out;
  transition: all 0.2s ease-out;
  width: 28px;
}

.emotion-0:hover .tag-remove {
  box-shadow: -4px 0 5px -1px rgba(48,48,48,0.2);
  opacity: 1;
  right: 0;
}

<div
  class="emotion-0 denali-tag"
  data-testid="tag"
>
  A tag
</div>
`;

exports[`Tag should render a small Tag 1`] = `
.emotion-0 {
  fill: rgba(48,48,48,0.6);
  cursor: pointer;
  vertical-align: text-bottom;
}

.emotion-1 {
  font-family: Helvetica,Arial,sans-serif;
  background: rgba(53,112,244,0.08);
  border-radius: 11px;
  border: 1px solid transparent;
  box-sizing: border-box;
  color: #303030;
  display: inline-block;
  font-size: 12px;
  height: 22px;
  line-height: 20px;
  margin: 3px;
  min-width: 50px;
  padding: 0 12px;
  position: relative;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: all 0.2s ease-out;
  transition: all 0.2s ease-out;
  vertical-align: top;
  white-space: nowrap;
}

.emotion-1:hover {
  background: #cbd9f9;
}

.emotion-1 .tag-remove {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #cbd9f9;
  border-radius: 0 11px 11px 0;
  box-shadow: none;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 22px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  opacity: 0;
  padding-right: 2px;
  position: absolute;
  right: -11px;
  top: -1px;
  -webkit-transition: all 0.2s ease-out;
  transition: all 0.2s ease-out;
  width: 22px;
}

.emotion-1:hover .tag-remove {
  box-shadow: -4px 0 5px -1px rgba(48,48,48,0.2);
  opacity: 1;
  right: 0;
}

<div
  class="emotion-1 denali-tag"
  data-testid="tag"
>
  A tag
  <div
    class="tag-remove"
  >
    <svg
      class="emotion-0"
      data-testid="icon"
      height="16px"
      viewBox="0 0 1024 1024"
      width="16px"
    >
      <title>
        x
      </title>
      <path
        d="M572.373 512l224.427-224.213c8.506-8.066 13.798-19.448 13.798-32.066 0-9.118-2.763-17.59-7.498-24.625l0.1 0.158c-7.858-10.565-20.305-17.335-34.333-17.335-11.534 0-21.999 4.577-29.678 12.013l0.011-0.011-227.2 225.707-226.133-225.707c-7.733-7.792-18.447-12.616-30.289-12.616-14.295 0-26.948 7.030-34.691 17.822l-0.087 0.128c-4.635 6.877-7.398 15.35-7.398 24.467 0 12.618 5.292 24 13.778 32.048l0.020 0.019 224.427 224.213-224.427 224.213c-8.506 8.066-13.798 19.448-13.798 32.066 0 9.118 2.763 17.59 7.498 24.625l-0.1-0.158c7.83 10.919 20.483 17.949 34.778 17.949 11.841 0 22.556-4.824 30.286-12.613l0.003-0.003 226.133-225.707 226.133 225.707c7.668 7.425 18.133 12.002 29.667 12.002 14.028 0 26.476-6.77 34.252-17.222l0.081-0.114c4.635-6.877 7.398-15.35 7.398-24.467 0-12.618-5.292-24-13.778-32.048l-0.020-0.019z"
      />
    </svg>
  </div>
</div>
`;
